cmd/routed-eni-cni-plugin/driver/driver.go

```diff
@@ -267,354 +271,358 @@ func WaitForAddressesToBeStable(ifName string, timeout time.Duration) error {
-	log.Infof("Added toContainer rule for %s", addrHostAddr.String())
+	// clean up any previous hostVeth ip rule recursively. (when pod with same name are recreated multiple times)
+	oldFromHostVethRule := n.netLink.NewRule()
+	oldFromHostVethRule.IifName = hostVethName
+	oldFromHostVethRule.Priority = vlanRulePriority
+	if err := netLinkRuleDelAll(n.netLink, oldFromHostVethRule); err != nil {
+		return errors.Wrapf(err, "SetupBranchENIPodNetwork: failed to delete hostVeth rule for %s", hostVethName)
+	}
 
-	// add from-pod rule, only need it when it is not primary ENI
-	if deviceNumber > 0 {
-		// To be backwards compatible, we will have to keep this off-by one setting
-		tableNumber := deviceNumber + 1
-		// add rule: 1536: from <podIP> use table <table>
-		err = addContainerRule(netLink, false, addrHostAddr, tableNumber)
-		if err != nil {
-			log.Errorf("Failed to add fromContainer rule for %s err: %v", addrHostAddr.String(), err)
-			return errors.Wrap(err, "add NS network: failed to add fromContainer rule")
+	rtTable := vlanID + 100
+	vlanLink, err := n.setupVlan(vlanID, eniMAC, subnetGW, parentIfIndex, rtTable, log)
+	if err != nil {
+		return errors.Wrapf(err, "SetupBranchENIPodNetwork: failed to setup vlan")
+	}
+
+	var containerAddr *net.IPNet
+	if v4Addr != nil {
+		containerAddr = v4Addr
+	} else if v6Addr != nil {
+		containerAddr = v6Addr
+	}
+	switch podSGEnforcingMode {
+	case sgpp.EnforcingModeStrict:
+		if err := n.setupIIFBasedContainerRouteRules(hostVeth, containerAddr, vlanLink, rtTable, log); err != nil {
+			return errors.Wrapf(err, "SetupBranchENIPodNetwork: unable to setup IIF based container routes and rules")
+		}
+	case sgpp.EnforcingModeStandard:
+		if err := n.setupIPBasedContainerRouteRules(hostVeth, containerAddr, rtTable, log); err != nil {
+			return errors.Wrapf(err, "SetupBranchENIPodNetwork: unable to setup IP based container routes and rules")
 		}
-		log.Infof("Added rule priority %d from %s table %d", fromContainerRulePriority, addrHostAddr.String(), tableNumber)
 	}
 	return nil
 }
 
+// TeardownBranchENIPodNetwork tears down the vlan and corresponding ip rules.
+func (n *linuxNetwork) TeardownBranchENIPodNetwork(containerAddr *net.IPNet, vlanID int, _ sgpp.EnforcingMode, log logger.Logger) error {
+	log.Debugf("TeardownBranchENIPodNetwork: containerAddr=%s, vlanID=%d", containerAddr.String(), vlanID)
+
+	if err := n.teardownVlan(vlanID, log); err != nil {
+		return errors.Wrapf(err, "TeardownBranchENIPodNetwork: failed to teardown vlan")
+	}
+
+	// to handle the migration between different enforcingMode, we try to clean up rules under both mode since the pod might be setup with a different mode.
+	rtTable := vlanID + 100
+	if err := n.teardownIIFBasedContainerRouteRules(rtTable, log); err != nil {
+		return errors.Wrapf(err, "TeardownBranchENIPodNetwork: unable to teardown IIF based container routes and rules")
+	}
+	if err := n.teardownIPBasedContainerRouteRules(containerAddr, rtTable, log); err != nil {
```

Comment:
Correct me if I am wrong, should we return error here? because if this is not migration then we might not have one of the rules to delete right?

Reply:
all the teardownIIFBasedContainerRouteRules and teardownIPBasedContainerRouteRules is designed to be safe to invoke if rules is not exists. 
so they can be called twice and return error safely :D